ddtrace/tracer/longrunner.go

```diff
@@ -0,0 +1,187 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+package tracer
+
+import (
+	"sync"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/internal"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+// TrackingExpirationLimit is the limit for how long to track a long-running span before no longer sending snapshots
+const TrackingExpirationLimit = 12 * time.Hour
+
+// Any span living longer than heartbeatInterval will have heartbeats sent every interval
+var heartbeatInterval time.Duration
+
+//TODO: is there a better performing design than this?
+type longrunner struct {
+	statsd statsdClient
+	// stopFunc is a fire exactly once method to ensure we don't try to stop more than once
+	stopFunc sync.Once
+	// done chan stops the long-running "work" goroutine
+	done chan struct{}
+	// mu protects the lower fields
+	mu sync.Mutex
+	// spans is a map of tracked spans to their "partial_version"
+	spans map[*span]int
+}
+
+func longrunningSpansEnabled(c *config) bool {
+	if c.longRunningEnabled && !c.agent.Info {
+		log.Warn("Long running span tracking requires a newer agent version than is connected")
+		return false
+	}
+	return c.longRunningEnabled
+}
+
+func limitHeartbeat(hb int64) time.Duration {
+	hbDur := time.Duration(hb)
+	const minInterval = 20 * time.Second
+	const maxInterval = (7 * time.Minute) + (30 * time.Second)
+	switch {
+	case hbDur < minInterval:
+		log.Warn("Long Running Span Configured interval too short, defaulting to minimum 20 seconds")
+		return minInterval
+	case hbDur > maxInterval:
+		log.Warn("Long Running Span Configured interval too long, defaulting to maximum 7.5 minutes")
+		return maxInterval
+	default:
+		return hbDur
+	}
+}
+
+// startLongrunner creates a long-running span tracker
+func startLongrunner(hbInterval int64, sd statsdClient) *longrunner {
+
+	heartbeatInterval = limitHeartbeat(hbInterval)
```

Comment:
Why do we need a globally declared variable here? If the answer is "for tests", then we don't need it :) Please make it a field.